packages/kit/src/pages.ts

```diff
@@ -34,3 +34,8 @@ export function extendRouteRules (route: string, rule: NitroRouteConfig, options
       : defu(opts.routeRules[route], rule)
   }
 }
+
+export function extendMiddleware (cb: (middleware: NuxtMiddleware[]) => void) {
+  const nuxt = useNuxt()
+  nuxt.hook('app:resolve', app => cb(app.middleware))
```

Comment:
It would be nice to:
- Support input argument being directly an array to inject
- Support returning new middleware from cb and push/normalize/dedup them into `app.middleware` <-- this would be much safer than just exposing an array and let developers push into it

Reply:
you want like this:

```ts
function extendMiddleware(newMiddleware: NuxtMiddleware[]): void;
function extendMiddleware(cb: (middleware: NuxtMiddleware[]) => NuxtMiddleware[]): void;
function extendMiddleware (input: any) {
  const nuxt = useNuxt()
  nuxt.hook('app:resolve', (app) => {
    if (Array.isArray(input)) {
      app.middleware = [...app.middleware, ...input]
    } else {
      app.middleware = input(app.middleware)
    }
  })
}
```

or add option argument for override middleware by name?

Reply:
This is my suggestion for the API:

```ts
function addRouteMiddleware (input: NuxtMiddleware | NuxtMiddleware[], options: { override: boolean }): void
```

Under the hood, if override is not set, then we will check existing middleware for any of the same name, and warn the user if so (and not replace the existing middleware).

Reply:
(Feel free to ping me on Discord if I can be of any help.)

Reply:
hi again

In `addRouteMiddleware` api, module can only add middleware. why not use [Adding middleware Dynamically](https://nuxt.com/docs/guide/directory-structure/middleware#adding-middleware-dynamically) helper.

but in `extendMiddleware` api, module can manage all middleware. like delete, sort and etc

an this moment, nuxt core team must decide which one to implement

tnx <3

Reply:
I would prefer `addRouteMiddleware`. Modules that need to do more advanced things can directly interact with the hook.

Reply:
you are right.
for advance change better to use hooks. 

if you are not comment or change this https://github.com/nuxt/nuxt/pull/18553#discussion_r1090448690 api, i can implement that.

Reply:
Yes, that would be perfect, thank you ❤️